test/asyncprogressqueueworker.js

```diff
@@ -4,60 +4,45 @@ const common = require('./common')
 const assert = require('assert');
 const os = require('os');
 
-test(require(`./build/${buildType}/binding.node`));
-test(require(`./build/${buildType}/binding_noexcept.node`));
+module.exports = test(require(`./build/${buildType}/binding.node`))
+  .then(() => test(require(`./build/${buildType}/binding_noexcept.node`)));
 
-function test({ asyncprogressqueueworker }) {
-  success(asyncprogressqueueworker);
-  fail(asyncprogressqueueworker);
-  cancel(asyncprogressqueueworker);
-  return;
+async function test({ asyncprogressqueueworker }) {
+  await success(asyncprogressqueueworker);
+  await fail(asyncprogressqueueworker);
 }
 
 function success(binding) {
-  const expected = [0, 1, 2, 3];
-  const actual = [];
-  const worker = binding.createWork(expected.length,
-    common.mustCall((err) => {
-      if (err) {
-        assert.fail(err);
-      }
-      // All queued items shall be invoked before complete callback.
-      assert.deepEqual(actual, expected);
-    }),
-    common.mustCall((_progress) => {
-      actual.push(_progress);
-    }, expected.length)
-  );
-  binding.queueWork(worker);
+  return new Promise((resolve, reject) => {
+    const expected = [0, 1, 2, 3];
+    const actual = [];
+    const worker = binding.createWork(expected.length,
+      common.mustCall((err) => {
+        if (err) {
+          reject(err);
+        } else {
+          // All queued items shall be invoked before complete callback.
+          assert.deepEqual(actual, expected);
+          resolve();
+        }
+      }),
+      common.mustCall((_progress) => {
+        actual.push(_progress);
+      }, expected.length)
+    );
+    binding.queueWork(worker);
+  });
 }
 
 function fail(binding) {
-  const worker = binding.createWork(-1,
-    common.mustCall((err) => {
-      assert.throws(() => { throw err }, /test error/)
-    }),
-    () => {
-      assert.fail('unexpected progress report');
-    }
-  );
-  binding.queueWork(worker);
-}
-
-function cancel(binding) {
```

Comment:
Had to remove this test, because create → queue → cancel is a racy operation. After queueing completes there is a small chance that by the time the next line – cancel – executes, the work will already have been queued. This was happening on

https://ci.nodejs.org/job/node-test-node-addon-api-new/2520/nodes=debian8-64/console
```
Error: Unknown failure
    at /home/iojs/build/workspace/node-test-node-addon-api-new/nodes/debian8-64/node-addon-api/test/asyncprogressqueueworker.js:67:13
    at new Promise (<anonymous>)
    at cancel (/home/iojs/build/workspace/node-test-node-addon-api-new/nodes/debian8-64/node-addon-api/test/asyncprogressqueueworker.js:52:10)
    at test (/home/iojs/build/workspace/node-test-node-addon-api-new/nodes/debian8-64/node-addon-api/test/asyncprogressqueueworker.js:13:9)
    at async /home/iojs/build/workspace/node-test-node-addon-api-new/nodes/debian8-64/node-addon-api/test/index.js:102:5
```

We have a similar test in core:

https://github.com/nodejs/node/blob/1f94b89309bcd56a2883cd1b0eb2db610251095e/test/node-api/test_async/test_async.cc#L140-L173

I can only conclude that it works in core because the stack is shallower and therefore the distance between queue and cancel is shorter and so it happens to work.